model.py

```diff
@@ -277,15 +298,26 @@ def _run(self, inputs, labels=None, device='CPU', device_ids=None):
                 feed[v.name] = labels[idx]
 
         endpoints = self._endpoints[self.mode]
-        fetch_list = endpoints['output'] + endpoints['loss']
-        num_output = len(endpoints['output'])
-        out = self._executor.run(
-            compiled_prog, feed=feed,
-            fetch_list=fetch_list)
         if self.mode == 'test':
-            return out[:num_output]
+            fetch_list = endpoints['output']
         else:
-            return out[:num_output], out[num_output:]
+            metric_list, metric_splits = flatten_list(endpoints['metric'])
+            fetch_list = endpoints['loss'] + metric_list
+            num_loss = len(endpoints['loss'])
+        rets = self._executor.run(
+            compiled_prog, feed=feed,
+            fetch_list=fetch_list,
+            return_numpy=False)
+        # LoDTensor cannot be fetch as numpy directly
+        rets = [np.array(v) for v in rets]
+        if self.mode == 'test':
+            return rets[:]
+        losses = rets[:num_loss]
+        metric_states = restore_flatten_list(rets[num_loss:], metric_splits)
+        metrics = []
+        for metric, state in zip(self.model._metrics, metric_states):
+            metrics.append(metric.update(*state))
+        return losses, metrics
```

Comment:
Maybe we should return metrics optionally  according to whether use metrics or not.

Reply:
Done, thanks!